### PR TITLE
Improved decorator types

### PR DESCRIPTION
**Pull request by thekiwi**
This PR adds the correct TypeScript decorator types, to improve type-safety and understandability of the codebase.

It also slightly refactors the GraphQL `@fieldResolver` decorator to improve the behaviour slightly:
- It no longer has the wrong signature (previously was the `ParameterDecorator` signature)
- It now has an interface that defines the actual metadata that we capture
- It now ensures the fieldResolver doesn't clash with another fieldResolver on the same class+property

I also added a couple of tests for verifying duplicate decorator functionality on the express decorators

### CODE REVIEW

File: packages/core/src/express/decorators/express.ts [hunk 1]
```diff
@@ -48,8 +48,8 @@ type Stage = 'before' | 'after';
  * @param middlewareFunction
  * @param stage
  */
-const middleware = (middlewareFunction, stage: Stage = 'before'): Function => {
-	return function(target: Record<string, any> | Function, methodName: string) {
+const middleware = (middlewareFunction, stage: Stage = 'before'): ClassDecorator & MethodDecorator => {
```
Comment (pierissimo) [inline packages/core/src/express/decorators/express.ts:51]: shouldn't this be `ClassDecorator | MethodDecorator`?

Comment (thekiwi) [reply packages/core/src/express/decorators/express.ts:51]: (For future reference...) According to [this StackOverflow comment](https://stackoverflow.com/questions/45306782/typescript-declaration-for-polymorphic-decorator#comment77857683_45306782) and [this similar code-change in Nest.js](https://github.com/nestjs/nest/commit/e8da23cd120e688ec97b5715184106a6c8061641#diff-2e475502839659c99ee212bbb1c78aaa8bdc15704e146433e56c31f9ac2d79c5R30) using `&` is the correct choice here.

File: packages/graphql/test/unit/decorators/index.test.ts [hunk 2]
```diff
@@ -24,4 +24,60 @@ describe('decorators', () => {
 			middlewares[2].middlewareFunction.should.be.equal(thirdMw);
 		});
 	});
+
+	describe('@graphResolver()', () => {
```
Comment (thekiwi) [inline packages/graphql/test/unit/decorators/index.test.ts:28]: typo 🤦🏻

**Review by pierissimo — APPROVED** [review]

**Review by Ugzuzg — APPROVED** [review]
